framework/Optimizers/Conjugate.py

```diff
@@ -0,0 +1,421 @@
+      direction = [var * factor for var in direction]
+
+    self.currentDirection = direction
+    return direction
+
+  def localEvaluateGradient(self, traj):
+    """
+      Local method to evaluate gradient.
+      @ In, traj, int, the trajectory id
+      @ Out, gradient, dict, dictionary containing gradient estimation. gradient should have the form {varName: gradEstimation}
+    """
+    gradient = {}
+    gi = {}
+    inVars = self.getOptVars()
+    opt = self.realizations[traj]['denoised']['opt'][0]
+    allGrads = self.realizations[traj]['denoised']['grad']
+    for g,pert in enumerate(allGrads):
+      varId = g % len(inVars)
+      var = inVars[varId]
+      lossDiff = mathUtils.diffWithInfinites(pert[self.objVar],opt[self.objVar])
+      # unlike SPSA, keep the loss diff magnitude so we get the exact right direction
+      if self.optType == 'max':
+        lossDiff *= -1.0
+      dh = pert[var] - opt[var]
+      if abs(dh) < 1e-15:
+        self.raiseADebug('Checking Var:',var)
+        self.raiseADebug('Opt point   :',opt)
+        self.raiseADebug('Grad point  :',pert)
+        self.raiseAnError(RuntimeError,'While calculating the gradArray a "dh" very close to zero was found for var:',var)
+      if gradient.get(var) == None:
+        gi[var] = 0
+        gradient[var] = np.atleast_1d(lossDiff / dh)
+      else:
+        gi[var] += 1
+        gradient[var] = (gradient[var] + np.atleast_1d(lossDiff / dh))/(gi[var]  + 1)
+    try:
+      self.counter['gradientHistory'][traj][1] = self.counter['gradientHistory'][traj][0]
+    except IndexError:
+      pass # don't have a history on the first pass
+    self.counter['gradientHistory'][traj][0] = gradient
+
+    return gradient
+
+  def _newOptPointAdd(self, gradient, traj):
+    """
+      This local method add a new opt point based on the gradient
+      @ In, gradient, dict, dictionary containing the gradient
+      @ In, traj, int, trajectory
+      @ Out, varKPlus, dict, new point that has been queued (or None if no new points should be run for this traj)
+    """
+    stepSize = self.counter['lastStepSize'][traj]
+    self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = {}
+    varK = dict((var,self.counter['recentOptHist'][traj][0][var]) for var in self.getOptVars())
+    varKPlus,modded = self._generateVarsUpdateConstrained(traj, stepSize, gradient, varK)
+    if len(self.optTrajLive) > 1 and self.counter['solutionUpdate'][traj] > 0:
+      removed = self._removeRedundantTraj(traj, varKPlus)
+    else:
+      removed = False
+    if removed:
+      return None
+    #if the new point was modified by the constraint, reset the step size
+    if modded:
+      self.counter['lastStepSize'][traj] = self.paramDict['initialStepSize']
+      self.raiseADebug('Resetting step size for trajectory',traj,'due to hitting constraints')
+    self.queueUpOptPointRuns(traj,varKPlus)
+    self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = varKPlus
+    return varKPlus
```

Comment:
see question above... How much?

Reply:
Changing step size in this step to be the line search result, and norm the gradient using the variable norm.